File: chartreuse/src/algorithms/ringbuffer.cc
/// @filename ringbuffer.cc
/// @brief Implementation of a simple resizable ringbuffer
/// @author gm
/// @copyright gm 2014
///
/// This file is part of Chartreuse
///
/// Chartreuse is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// Chartreuse is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with Chartreuse.  If not, see <http://www.gnu.org/licenses/>.

// std::fill, std::copy, std::min, std::max
#include <algorithm>

#include "chartreuse/src/algorithms/ringbuffer.h"
#include "chartreuse/src/common.h"

namespace chartreuse {
namespace algorithms {

RingBuffer::RingBuffer(const std::size_t capacity)
    : data_(nullptr),
      capacity_(capacity),
      size_(0),
      writing_position_(0),
      reading_position_(0) {
  CHARTREUSE_ASSERT(capacity > 0);
  data_ = static_cast<float*>(new float[capacity_]);
  std::fill_n(&data_[0], capacity_, 0.0f);
}

RingBuffer::~RingBuffer() {
  delete[] data_;
  data_ = nullptr;
}

void RingBuffer::PopOverlapped(float* dest,
                               const std::size_t count,
                               const unsigned int overlap) {
  CHARTREUSE_ASSERT(IsGood());
  CHARTREUSE_ASSERT(overlap > 0);

  // Is zero padding required ?
  // Using "ints" here cause count may be < size, or the opposite
  const std::size_t zeropadding_count(static_cast<std::size_t>(
    (std::max(static_cast<int>(count - size_), 0))));
  // Actual elements count to be copied
  const std::size_t copy_count(count - zeropadding_count);

  // Length of the "right" part: from reading cursor to the buffer end
  const std::size_t right_part_size(std::min(capacity_ - reading_position_,
                                     copy_count));
  // Length of the "left" part: from the buffer beginning
  // to the last element to be copied
  const std::size_t left_part_size(copy_count - right_part_size);


  //  Copy the first part
  std::copy_n(&data_[reading_position_], right_part_size, &dest[0]);
  //  Copy the second part
  std::copy_n(&data_[0], left_part_size, &dest[right_part_size]);

  reading_position_ += copy_count / overlap;
  reading_position_ = reading_position_ % capacity_;

  size_ -= copy_count / overlap;

  // Zero-padding
  std::fill_n(&dest[copy_count],
              zeropadding_count,
              0.0f);
}

void RingBuffer::Pop(float* dest, const std::size_t count) {
  return PopOverlapped(dest, count, 1);
}

void RingBuffer::Push(const float* const src, const std::size_t count) {
  CHARTREUSE_ASSERT(IsGood());
  CHARTREUSE_ASSERT(count <= Capacity() - Size());
  // Length of the "right" part: from writing cursor to the buffer end
  const std::size_t right_part_size(std::min(capacity_ - writing_position_,
                                              count));
  // Length of the "left" part: from the buffer beginning
  // to the last element to be pushed
  const std::size_t left_part_size(count - right_part_size);

  //  Copy the first part
  std::copy_n(&src[0], right_part_size, &data_[writing_position_]);
  //  Copy the second part
  std::copy_n(&src[right_part_size], left_part_size, &data_[0]);

  writing_position_ += count;
  writing_position_ = writing_position_ % capacity_;

  size_ += count;
}

void RingBuffer::Fill(const float value, const std::size_t count) {
  CHARTREUSE_ASSERT(IsGood());
  CHARTREUSE_ASSERT(count > 0);
  CHARTREUSE_ASSERT(count <= Capacity() - Size());
  // Length of the "right" part: from writing cursor to the buffer end
  const std::size_t right_part_size(std::min(capacity_ - writing_position_,
                                              count));
  // Length of the "left" part: from the buffer beginning
  // to the last element to be pushed
  const std::size_t left_part_size(count - right_part_size);

  // Filling the first part
  std::fill_n(&data_[writing_position_], right_part_size, value);
  // Fill the second part
  std::fill_n(&data_[0], left_part_size, value);

  writing_position_ += count;
  writing_position_ = writing_position_ % capacity_;

  size_ += count;
}

void RingBuffer::Clear(void) {
  writing_position_ = 0;
  reading_position_ = 0;
  size_ = 0;
  if (IsGood()) {
    std::fill_n(&data_[0], Capacity(), 0.0f);
  }
}

bool RingBuffer::IsGood(void) const {
  return data_ != nullptr;
}

std::size_t RingBuffer::Capacity(void) const {
  CHARTREUSE_ASSERT(IsGood());
  return capacity_;
}

std::size_t RingBuffer::Size(void) const {
  CHARTREUSE_ASSERT(IsGood());
  return size_;
}

}  // namespace algorithms
}  // namespace chartreuse

File: scripts/ringbuffer.py
#!/usr/bin/env python
'''
@file ringbuffer.py
@brief A simple ringbuffer implementation
@author gm
@copyright gm 2014

This file is part of Chartreuse

Chartreuse is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Chartreuse is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Chartreuse.  If not, see <http://www.gnu.org/licenses/>.
'''
import numpy
import pylab

class RingBuffer(object):
    '''
    A simple non-resizable 1D ringbuffer
    '''
    def __init__(self, capacity, overlap):
        self._data = numpy.zeros(capacity)
        self.capacity = capacity
        self.size = 0
        self._read_cursor = 0
        self._write_cursor = 0
        self._overlap = overlap

    def IsFull(self):
        return self.size == self.capacity

    def IsEmpty(self):
        return self.size == 0

    def Push(self, data):
        data_length = len(data)
        if data_length > self.capacity:
            raise Exception("Too much input data!")

        # TODO(gm): check out numpy.roll
        right_part_length = min(data_length, self.capacity -  self._write_cursor)
        left_part_length = data_length - right_part_length
        self._data[self._write_cursor:self._write_cursor + right_part_length] = data[0:right_part_length]
        self._data[0:left_part_length] = data[right_part_length:right_part_length + left_part_length]

        self._write_cursor += data_length
        self._write_cursor = self._write_cursor % self.capacity
        self.size += data_length

    def Fill(self, value, data_length):
        return self.Push(value * numpy.zeros(data_length))

    def Pop(self, data_length):
        # Zero-padding may occur!
        actual_copy_count = min(self.size, data_length)

        out = numpy.zeros(data_length)
        # TODO(gm): check out numpy.roll
        right_part_length = min(actual_copy_count, self.capacity -  self._read_cursor)
        left_part_length = actual_copy_count - right_part_length
        out[0:right_part_length] = self._data[self._read_cursor:self._read_cursor + right_part_length]
        out[right_part_length:right_part_length + left_part_length] = self._data[0:left_part_length]

        self._read_cursor += data_length
        self._read_cursor = self._read_cursor % self.capacity
        self.size -= data_length

        # Zero-padding
        if actual_copy_count < data_length:
            out[actual_copy_count:data_length] = numpy.zeros(data_length - actual_copy_count)

        return out

    def PopOverlapped(self, data_length):
        # Zero-padding may occur!
        actual_copy_count = min(self.size, data_length)

        out = numpy.zeros(data_length)
        # TODO(gm): check out numpy.roll
        right_part_length = min(actual_copy_count, self.capacity -  self._read_cursor)
        left_part_length = actual_copy_count - right_part_length
        out[0:right_part_length] = self._data[self._read_cursor:self._read_cursor + right_part_length]
        out[right_part_length:right_part_length + left_part_length] = self._data[0:left_part_length]

        self._read_cursor += data_length / self._overlap
        self._read_cursor = self._read_cursor % self.capacity
        self.size -= data_length / self._overlap

        # Zero-padding
        if actual_copy_count < data_length:
            out[actual_copy_count:data_length] = numpy.zeros(data_length - actual_copy_count)

        return out

if __name__ == "__main__":
    import utilities

    frame_length = 480
    length = 9600
    overlap = 3

    out_data = numpy.zeros(length)

    in_data = numpy.random.rand(length)

    ringbuffer = RingBuffer(length, overlap)

    current_in_idx = 0
    while not ringbuffer.IsFull():
        ringbuffer.Push(in_data[current_in_idx * frame_length: (current_in_idx + 1) * frame_length])
        current_in_idx += 1

    current_out_idx = 0
    while not ringbuffer.IsEmpty():
        out_data[current_out_idx * frame_length: (current_out_idx + 1) * frame_length] = ringbuffer.Pop(frame_length)
        current_out_idx += 1

    print(utilities.PrintMetadata(utilities.GetMetadata(out_data[0:length] - in_data)))

    pylab.plot(in_data[0:length], label = "in")
    pylab.plot(out_data[0:length], label = "out")

    pylab.legend()
    pylab.show()
